lookout/style/format/features.py

```diff
@@ -225,17 +247,95 @@ def _classify_vnodes(self, nodes: Iterable[VirtualNode]) -> List[VirtualNode]:
+        :param vnode_index: the column at which we'll start writing the features is
+                            vnode_index * n_features.
+        """
+        role_index = -1
+        if vnode.node:
+            role = vnode.node.internal_type
+            if role in self.roles.ROLE_INDEX:
+                role_index = self.roles.ROLE_INDEX[role]
+        col = vnode_index * len(self.feature_names)
+        features[row, col:col + 3] = vnode.start
+        features[row, col + 3:col + 6] = vnode.end
+        features[row, col + 6] = role_index
+
+    def _inplace_write_vnodes_features(self, vnodes: Sequence[VirtualNode],
+                                       parents: Mapping[int, bblfsh.Node], parents_depth: int,
+                                       siblings_window: int, line_offset: int, X: numpy.ndarray
+                                       ) -> None:
+        """
+        Given a sequence of `VirtualNode`-s and relevant info, compute the input matrix and label
+        vector.
+
+        :param vnodes: sequence of `VirtualNode`-s
+        :param parents: dictionnary of node id to parent node
+        :param parents_depth: number of parents to use in the features
+        :param siblings_window: number of siblings to use, both at the left and right of the node
+        :param line_offset: at which line of the input ndarrays should we start writing
+        :param X: features matrix
+        """
+        last_left_parent = None
+        i = 0
+        for vnode in vnodes:
+
+            # compute current vnode parent. If both left and right closest parent are the same then
+            # we use it else we use no parent feature (set them to -1 later on)
+            if vnode.node:
+                last_left_parent = parents[id(vnode.node)]
+            if not vnode.y:
+                continue
+            next_right_parent = None
+            for future_vnode in vnodes[i + 1:]:
+                if future_vnode.node:
+                    next_right_parent = parents[id(future_vnode.node)]
+                    break
+            parent = last_left_parent if last_left_parent is next_right_parent else None
+
+            # compute how many dummy features we'll need to account for the lack of left and right
+            # siblings and the lack of parents
+            n_dummies_left = abs(min(i - siblings_window, 0))
+            parents_list = []
+            if parent:
+                current_vnode = vnode
+                for j in range(parents_depth):
+                    node_id = id(current_vnode)
+                    if node_id in parents:
+                        parent = parents[node_id]
+                        parents_list.append(parent)
+                        current_vnode = parent
+                    else:
+                        break
+            self._inplace_write_vnode_features(X, vnode, i + line_offset, 0)
+            for j, left_vnode in enumerate(vnodes[max(0, i - siblings_window):i]):
```

Comment:
could you please put a comment here...

what do we write for each node. Window left, window right, parent... `parents_list`?